models/repo/avatar.go

```diff
@@ -84,7 +84,13 @@ func (repo *Repository) relAvatarLink(ctx context.Context) string {
 	return setting.AppSubURL + "/repo-avatars/" + url.PathEscape(repo.Avatar)
 }
 
-// AvatarLink returns the full avatar url with http host. TODO: refactor it to a relative URL, but it is still used in API response at the moment
+// AvatarLink returns the full avatar url with http host or the empty string if the repo doesn't have an avatar.
+//
+// TODO: refactor it to a relative URL, but it is still used in API response at the moment
 func (repo *Repository) AvatarLink(ctx context.Context) string {
-	return httplib.MakeAbsoluteURL(ctx, repo.relAvatarLink(ctx))
```

Comment:
I am wondering is we should instead perform this empty string check in `MakeAbsoluteURL()`?

Reply:
FYI the main culprit is `url.Parse()` which exists within this function call chain: `MakeAbsoluteURL()` -> `IsRelativeURL()` -> `url.Parse()`. It accepts empty strings.

Reply:
There are only two uses of `MakeAbsoluteURL`, so I suggest doing the validation there and add a `err` return value for invalid inputs like a empty string.

```
models/repo/avatar.go:89:	return httplib.MakeAbsoluteURL(ctx, repo.relAvatarLink(ctx))
models/user/avatar.go:94:	return httplib.MakeAbsoluteURL(ctx, u.AvatarLinkWithSize(ctx, 0))

Reply:
Thought it raises the question, is empty string really invalid? One could see it being valid and the function just returning the current url origin.

Reply:
It's fine for "user avatar" because it could never be empty IIRC.

And empty string is valid URL. For example: `href=""` or `href="?k=v"`.

So the check should be done in `AvatarLink` function.

Reply:
> Thought it raises the question, is empty string really invalid? One could see it being valid and the function just returning the current url origin. In this case it makes sense to perform the validation at the caller site.

> So the check should be done in AvatarLink function.

I agree. Returning an empty URL string appears to be a pattern that is followed in the API when dealing with missing/non-existent URLs and I don't think we should force this on library code (since in the future there maybe callers that depend on the current behavior).